sherpa/ui/utils.py

```diff
@@ -224,6 +224,111 @@ def _remove_obj_from_main(name):
             pass
 
 
+def set_dep(data, val):
+    """Set the dependent axis.
+
+    Parameters
+    ----------
+    data : sherpa.data.Data instance
+    val : scalar or array_like
+        If a scalar then use the same value for all elements.
+
+    """
+
+    if numpy.iterable(val):
+        dep = numpy.asarray(val, SherpaFloat)
+    else:
+        val = SherpaFloat(val)
+        dep = numpy.array([val] * data.size)
+
+    data.dep = dep
+
+
+def set_error(data, field, val, fractional=False):
+    """Set the error field.
+
+    Parameters
+    ----------
+    data : sherpa.data.Data instance
+    field : {"staterror", "syserror"}
+    val : None, a scalar, or array_like
+        The statistical error.
+    fractional : bool, optional
+        If `val` is a scalar then set errors to
+        `val * data.y`.
+
+    Notes
+    -----
+    The `val` field, when an array, must match the data.size
+    field, even for grouped PHA data.
+
+    """
+
+    if val is None:
+        err = None
+
+    elif numpy.iterable(val):
+        err = numpy.asarray(val, SherpaFloat)
+
+    else:
+        val = SherpaFloat(val)
+        if sherpa.utils.bool_cast(fractional):
+            err = val * data.get_dep()
+        else:
+            err = numpy.array([val] * data.size)
+
+    setattr(data, field, err)
+
+
+def set_filter(data, val, ignore=False):
+    """Set the filter field.
+
+    Parameters
+    ----------
+    data : sherpa.data.Data instance
+    val : array_like
+        The filter array (bool-like).
+    ignore : bool, optional
+        If set the filter should be inverted.
+
+    Notes
+    -----
+    The `val` field should match the mask array, that is the
+    "effective" size of the data. This is only different from
+    data.size when the data is grouped (and at least one group
+    contains multiple channels).
+
+    """
+
+    val = numpy.asarray(val, dtype=numpy.bool_)
+    nval = len(val)
+
+    # Note that we do not use data.size as the check, as that fails
+    # for grouped PHA data. Is there some way to have an "effective
+    # size" value?
+    #
+    nexp = len(data.get_y(False))
+    if numpy.iterable(data.mask):
+        if nexp != nval:
+            raise sherpa.utils.err.DataErr('mismatchn', 'data', 'filter',
+                                           nexp, nval)
+
+        if not ignore:
+            data.mask |= val
+        else:
+            data.mask &= ~val
+
+        return
+
+    # The mask attribute checks the length, so we do not have to here.
+    #
+    if not ignore:
+        data.mask = val
+    else:
+        assert False, "D"
```

Comment:
?

Reply:
Hmm, that looks like a debugging statement. I shall have to investigate (and probably just remove the `else` clause).

Reply:
Going to remove it as part of the next rebase

Reply:
It's been removed, but leaving the conversation only for now.